drv/gimlet-hf-server/src/main.rs

```diff
@@ -180,6 +269,190 @@ impl ServerImpl {
+        set_and_check_write_enable(qspi)?;
+        qspi.page_program(addr, data);
+        poll_for_write_complete(qspi, None);
+        Ok(())
+    }
+
+    fn set_dev(&mut self, state: HfDevSelect) -> Result<(), HfError> {
+        // Return early if the dev select pin is missing
+        let dev_select_pin = self.dev_select_pin.ok_or(HfError::NoDevSelect)?;
+
+        self.check_muxed_to_sp()?;
+
+        let sys = sys_api::Sys::from(SYS.get_task_id());
+        match state {
+            HfDevSelect::Flash0 => sys.gpio_reset(dev_select_pin),
+            HfDevSelect::Flash1 => sys.gpio_set(dev_select_pin),
+        }
+
+        self.dev_state = state;
+        Ok(())
+    }
+
+    fn get_raw_persistent_data(
+        &mut self,
+    ) -> Result<RawPersistentData, HfError> {
+        self.check_muxed_to_sp()?;
+        let best = if self.dev_select_pin.is_some() {
+            let prev_slot = self.dev_state;
+
+            // At this point, all of the ways that `set_dev` could fail have
+            // been checked: we already called both `check_muxed_to_sp` and that
+            // `dev_select_pin` is `Some(...)`, so we can unwrap returns from
+            // `self.set_dev(...)`.
+
+            // Look at the inactive slot first
+            self.set_dev(!prev_slot).unwrap();
+            let (a, _) = self.persistent_data_scan();
+
+            // Then switch back to our current slot, so that the resulting state
+            // is unchanged.
+            self.set_dev(prev_slot).unwrap();
+            let (b, _) = self.persistent_data_scan();
+
+            match (a, b) {
+                (Some(a), Some(b)) => {
+                    if a.epoch > b.epoch {
+                        Some(a)
+                    } else {
+                        Some(b)
+                    }
+                }
+                (Some(_), None) => a,
+                (None, Some(_)) => b,
+                (None, None) => None,
+            }
+        } else {
+            // The single-chip case is easy:
+            self.persistent_data_scan().0
+        };
+
+        best.ok_or(HfError::NoPersistentData)
+    }
+
+    /// Scans the currently active flash IC for persistent data
+    ///
+    /// Returns a tuple containing
+    /// - newest persistent data record
+    /// - address of first empty slot
+    fn persistent_data_scan(&self) -> (Option<RawPersistentData>, Option<u32>) {
+        let mut best: Option<RawPersistentData> = None;
+        let mut empty_slot: Option<u32> = None;
+        for i in 0..SECTOR_SIZE_BYTES / HF_PERSISTENT_DATA_STRIDE {
+            let addr = (i * HF_PERSISTENT_DATA_STRIDE) as u32;
+            let mut data = RawPersistentData::new_zeroed();
+            self.qspi.read_memory(addr, data.as_bytes_mut());
+            if data.is_valid()
+                && best.map(|b| b.epoch < data.epoch).unwrap_or(true)
+            {
+                best = Some(data)
+            }
+            if empty_slot.is_none()
```

Comment:
This might be slightly clearer as an `else if` - if `data.is_valid()`, we know that some of the bytes are not `0xff` already.

Reply:
Hmmm, that's logically correct, but I feel like it's more clear as-is, since the two variables that we're populating (`best` and `empty_slot`) are philosophically distinct.

Reply:
Huh, I thought of them more closely related - it's not possible for one slot to be both `best` and `empty` at the same time (hence thinking "else if" would make that clear). But this is not a hill I'm dying on - either is fine!